src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.util.UnitFormatter;
+
+abstract class AbstractMessageLog<T extends MessageLog>
+        extends CompletableFuture<T> implements MessageLog, MessageLogBuilder {
+
+    private boolean startTimeNanosSet;
+    private long startTimeNanos;
+    private long contentLength;
+    private long endTimeNanos;
+    private List<Object> attachments;
+    private Throwable cause;
+
+    boolean start0() {
+        if (isDone()) {
+            return false;
+        }
+
+        if (!startTimeNanosSet) {
+            startTimeNanos = System.nanoTime();
+            startTimeNanosSet = true;
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean isStarted() {
+        return startTimeNanosSet;
+    }
+
+    @Override
+    public long startTimeNanos() {
+        return startTimeNanos;
+    }
+
+    @Override
+    public void increaseContentLength(long deltaBytes) {
+        if (deltaBytes < 0) {
+            throw new IllegalArgumentException("deltaBytes: " + deltaBytes + " (expected: >= 0)");
+        }
+        if (isDone()) {
+            return;
+        }
+
+        contentLength += deltaBytes;
+    }
+
+    @Override
+    public void contentLength(long contentLength) {
+        if (contentLength < 0) {
+            throw new IllegalArgumentException("contentLength: " + contentLength + " (expected: >= 0)");
+        }
+        if (isDone()) {
+            return;
+        }
+
+        this.contentLength = contentLength;
+    }
+
+    @Override
+    public long contentLength() {
+        return contentLength;
+    }
+
+    @Override
+    public void attach(Object attachment) {
```

Comment:
seems multiple `HttpHeaders` can be attached, but `attachment(type)` allows only one instance will be returned. I feel this attachment design is hard to understand...
as no-brain suggestion, how about using `AttributeMap` instead?